hivemq-edge/src/frontend/src/extensions/datahub/designer/operation/OperationPanel.tsx

```diff
@@ -31,23 +35,53 @@ export const OperationPanel: FC<PanelProps> = ({ selectedNode, onFormSubmit }) =
     return adapterNode?.data
   }, [edges, nodes, selectedNode])
 
+  const pipelineIds = useMemo(() => {
+    const operationNode = nodes.find((e) => e.id === selectedNode) as Node<OperationData> | undefined
+    if (!operationNode?.data) return null
+
+    const getAllParents = (node: Node, visited = new Set<Node>()): Set<Node> => {
```

Comment:
suggestion (non-blocking): add a unit test for `getAllParents()` (which would mean moving it to some utility functions file) for line & branch coverage.

Reply:
+1 would also reduce the complexity of the function and it's redeclaration on state change since it's pure.